File: NeetCode/Trees/Valid_Binary_Search_Tree/valid__binary__search__tree.py
from typing import Optional, List

from common.treenode import TreeNode, deserialize, printTree, null


class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def dfs(node, interval):
            if node is None:
                return True
            if node and (interval[0] >= node.val or interval[1] <= node.val):
                return False
            checkLeft = dfs(node.left, [interval[0], node.val])
            checkRight = dfs(node.right, [node.val, interval[1]])
            return checkLeft and checkRight

        return dfs(root, [float("-inf"), float("inf")])


if __name__ == "__main__":
    sol = Solution()
    root = deserialize([5, 4, 6, null, null, 3, 7])
    print(sol.isValidBST(root))
    # root = deserialize([1, 2, 3])
    # print(sol.isValidBST(root))
    print("Running Solution...")

File: NeetCode/Trees/Valid_Binary_Search_Tree/Valid_Binary_Search_Tree.cpp
// Valid_Binary_Search_Tree.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
class Solution
{
public:
    bool isValidBST(TreeNode *root)
    {
        return dfs(root, LLONG_MIN, LLONG_MAX);
    }

    bool dfs(TreeNode *node, long long minimum, long long maximum)
    {
        if (!node)
        {
            return true;
        }
        if (node->val >= minimum || node->val >= maximum)
        {
            return false;
        }
        return dfs(node->left, minimum, node->val) && dfs(node->right, node->val, maximum);
    }
};
